workspaces/diff-engine/src/commands/mod.rs

```diff
@@ -63,17 +68,105 @@ impl std::error::Error for SpecCommandError {}
 // CommandContext
 // --------------
 
-#[derive(Deserialize, Debug, PartialEq)]
+#[derive(Clone, Deserialize, Debug, PartialEq)]
 #[serde(rename_all = "camelCase")]
 pub struct CommandContext {
   pub client_id: String,
   pub client_session_id: String,
   pub client_command_batch_id: String,
+  pub created_at: DateTime<Utc>,
+}
+
+impl CommandContext {
+  pub fn new(
+    batch_id: String,
+    client_id: String,
+    client_session_id: String,
+    created_at: DateTime<impl TimeZone>,
+  ) -> Self {
+    Self {
+      client_id,
+      client_command_batch_id: batch_id,
+      client_session_id,
+      created_at: created_at.with_timezone(&Utc),
+    }
+  }
+}
+
+impl Default for CommandContext {
+  fn default() -> Self {
+    Self {
+      client_id: String::from("anonymous"),
+      client_session_id: String::from("unknown-session"),
+      client_command_batch_id: String::from("unknown-batch"),
+      created_at: Utc.timestamp(0, 0),
+    }
+  }
+}
+
+// SpecCommandHandler
+// ------------------
+
+#[derive(Debug, Default)]
+pub struct SpecCommandHandler {
+  command_context: CommandContext,
+  spec_projection: SpecProjection,
+}
+
+impl SpecCommandHandler {
+  pub fn new(command_context: CommandContext, spec_projection: SpecProjection) -> Self {
+    Self {
+      command_context,
+      spec_projection,
+    }
+  }
+
+  pub fn with_command_context(&mut self, command_context: CommandContext) {
+    self.command_context = command_context
+  }
+}
+
+impl Aggregate for SpecCommandHandler {
+  fn aggregate_type() -> &'static str {
+    "spec-command-handler"
+  }
+}
+
+impl AggregateEvent<SpecCommandHandler> for SpecEvent {
+  fn apply_to(self, command_handler: &mut SpecCommandHandler) {
+    command_handler.spec_projection.apply(self)
+  }
+}
+
+impl<I> From<I> for SpecCommandHandler
+where
+  I: IntoIterator,
+  I::Item: AggregateEvent<SpecProjection>,
+{
+  fn from(events: I) -> Self {
+    let spec_projection = SpecProjection::from(events);
+    Self::new(CommandContext::default(), spec_projection)
+  }
 }
 
 // Command handling
 // ----------------
 
+impl AggregateCommand<SpecCommandHandler> for SpecCommand {
+  type Error = SpecCommandError;
+  type Event = SpecEvent;
+  type Events = Vec<SpecEvent>;
+
+  fn execute_on(self, command_handler: &SpecCommandHandler) -> Result<Self::Events, Self::Error> {
+    let mut events = command_handler.spec_projection.execute(self)?;
+    let event_context = EventContext::from(command_handler.command_context.clone());
+    for event in &mut events {
+      event.with_event_context(event_context.clone());
+    }
+    Ok(events)
+  }
+}
+
```

Comment:
Ended up only wrapping this on a surface-level right now and assigning the events with context once they have been created. I took at approach as I ran into the issue of having to propagate that command context to nested handlers and projections, which would have taken way more refactoring for a use case that we're not actually hitting at the moment (using command context in command handling).

Once we do require using the command context in the command handling logic, this should give us a good starting off point, regardless. Then, we'd also have something to actually test and measure whether how we've implemented it makes sense.